{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eae35c4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import required libraries\n",
    "import pandas\n",
    "import matplotlib.pyplot as plt\n",
    "from Bio.Blast import NCBIWWW, NCBIXML\n",
    "import re\n",
    "from Bio import Entrez\n",
    "\n",
    "# Define function to run BLAST search and save results to a file\n",
    "def run_blast(query_seq, database_name, entrez_query, nucl_penalty=-1, nucl_reward=1, word_size=7):\n",
    "    \"\"\"Performs a BLAST search and saves the results to a file.\"\"\"\n",
    "    with open(query_seq) as f:\n",
    "        sequence_data = f.read()\n",
    "        \n",
    "# Perform BLAST search using NCBIWWW.qblast function and save results to a file\n",
    "\n",
    "    result_handle = NCBIWWW.qblast(\n",
    "        program=\"blastn\",\n",
    "        database=database_name,\n",
    "        entrez_query=entrez_query,\n",
    "        sequence=sequence_data,\n",
    "        nucl_penalty=nucl_penalty,\n",
    "        nucl_reward=nucl_reward,\n",
    "        word_size=word_size\n",
    "    )\n",
    "\n",
    "    with open('results_2.xml', 'w') as f:\n",
    "        f.write(result_handle.read())\n",
    "\n",
    "# Define function to remove a word from an XML file and save the updated file    \n",
    "def remove_word_from_xml(xml_file, word, output_file):\n",
    "    \"\"\"Removes a word from an XML file and saves the updated file.\"\"\"\n",
    "    with open(xml_file) as f:\n",
    "        content = f.read()\n",
    "    content = re.sub(word, '', content)\n",
    "    with open(output_file, 'w') as f:\n",
    "        f.write(content)\n",
    "\n",
    "# Define function to parse BLAST results and save them to a file\n",
    "def parse_results(xml_file, output_file, evalue_thresh):\n",
    "    \"\"\"Parses the BLAST results and saves them to a file.\"\"\"\n",
    "    with open(xml_file) as f:\n",
    "        content = f.read()\n",
    "    with open(output_file, 'w') as f:\n",
    "        for record in NCBIXML.parse(open(xml_file)):\n",
    "            if not record.alignments:\n",
    "                continue\n",
    "            for align in record.alignments:\n",
    "                for hsp in align.hsps:\n",
    "                    if hsp.expect < evalue_thresh:\n",
    "                        hit_id = align.hit_id.split('|')\n",
    "                        hit_def = align.hit_def.split(' ')\n",
    "                        species = f\"{hit_def[0]} {hit_def[1]}\"\n",
    "                        uid_new = hit_id[3]\n",
    "                        f.write(f\"{uid_new}\\t{hsp.sbjct_start}\\t{hsp.sbjct_end}\\t{species}\\t{hsp.expect}\\t{hsp.score}\\t{hsp.align_length}\\n\")\n",
    "\n",
    "# Define function to filter results based on score and plot the score distribution\n",
    "def filter_score():\n",
    "    # Read data from input file using pandas.read_csv function\n",
    "    df = pandas.read_csv(\"output_2.txt\", sep='\\t',skiprows= [0],names=['ID', 'Start','Stop','Species','E-value','Score','Align_length'])\n",
    "    data = df['Score']\n",
    "    # Calculate mean, standard deviation, and upper limit based on standard deviation of the score distribution\n",
    "    mean = df['Score'].mean()\n",
    "    std = df['Score'].std()\n",
    "    limit =std\n",
    "    # Determine minimum and maximum score values for plotting purposes\n",
    "    min_value = min(data)\n",
    "    max_value = max(data)\n",
    "    # Plot the score distribution using matplotlib.pyplot.scatter and matplotlib\n",
    "    plt.title(\"Score_values\")\n",
    "    plt.ylim(min_value - 100, max_value + 100)\n",
    "    plt.scatter(x=df.index, y=df['Score'])\n",
    "    plt.hlines(y= std, xmin=0, xmax=len(data),colors='g')\n",
    "    plt.show()\n",
    "    \n",
    "    print(limit)\n",
    "    \n",
    "    df_new = df[df['Score'] > limit]\n",
    "    print(df_new)\n",
    "\n",
    "    df_new.to_csv('final_op2.txt', sep=\"\\t\",index= False,header= False)\n",
    "\n",
    "#This function reads the input file, extracts the ID, start and end positions of each sequence, fetches the corresponding sequences using the Entrez API, and writes the output to a file in FASTA format.\n",
    "def flank(bp, input_file, output_file):\n",
    "    uid = []\n",
    "    start = []\n",
    "    end = []\n",
    "    with open(input_file) as f:\n",
    "        rows = (line.split('\\t') for line in f)\n",
    "        for row in rows:\n",
    "            uid.append(row[0])\n",
    "            start.append(int(row[1]))\n",
    "            end.append(int(row[2]))\n",
    "        \n",
    "    # Running efetch and returning the output in FASTA format\n",
    "    Entrez.email = \"ishaguru64@gmail.com\"\n",
    "    with open(output_file, \"w\") as bed:\n",
    "        for i in range(len(uid)):\n",
    "            try:\n",
    "                handle_new = Entrez.efetch(db=\"nuccore\", id=uid[i], rettype='fasta', retmode='txt', seq_start=start[i]-bp, seq_stop=end[i]+bp)\n",
    "                print(handle_new.read(), file=bed)\n",
    "            except Exception as e:\n",
    "                print(f\"Error fetching sequence {uid[i]}: {str(e)}\")\n",
    "\n",
    "    \n",
    "#calling main function   \n",
    "if __name__ == \"__main__\":\n",
    "    run_blast(\"agam_crm.txt\" ,\"refseq_genomic\" ,\"txid43816[ORGN]\")\n",
    "    remove_word_from_xml(\"results_2.xml\", \"CREATE_VIEW\",\"file_2.xml\")\n",
    "    parse_results(\"file_2.xml\", \"output_2.txt\", 0.05)\n",
    "    filter_score()\n",
    "    flank(10, \"final_op2.txt\", \"agam_file\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92359e2d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
